source/schema/avro.go

```diff
@@ -0,0 +1,161 @@
+		"float8":  avro.NewPrimitiveSchema(avro.Double, nil),
+		"int8":    avro.NewPrimitiveSchema(avro.Long, nil),
+		"int4":    avro.NewPrimitiveSchema(avro.Int, nil),
+		"int2":    avro.NewPrimitiveSchema(avro.Int, nil),
+		"text":    avro.NewPrimitiveSchema(avro.String, nil),
+		"varchar": avro.NewPrimitiveSchema(avro.String, nil),
+		"timestamptz": avro.NewPrimitiveSchema(
+			avro.Long,
+			avro.NewPrimitiveLogicalSchema(avro.TimestampMicros),
+		),
+		"timestamp": avro.NewPrimitiveSchema(
+			avro.Long,
+			avro.NewPrimitiveLogicalSchema(avro.LocalTimestampMicros),
+		),
+		"date": avro.NewPrimitiveSchema(
+			avro.Int,
+			avro.NewPrimitiveLogicalSchema(avro.Date),
+		),
+		"uuid": avro.NewPrimitiveSchema(
+			avro.String,
+			avro.NewPrimitiveLogicalSchema(avro.UUID),
+		),
+	},
+}
+
+type avroExtractor struct {
+	pgMap   *pgtype.Map
+	avroMap map[string]*avro.PrimitiveSchema
+}
+
+func (a avroExtractor) ExtractLogrepl(rel *pglogrepl.RelationMessage, row *pglogrepl.TupleData) (avro.Schema, error) {
+	var (
+		fields []pgconn.FieldDescription
+		values []any
+	)
+
+	for i, tuple := range row.Columns {
+		fields = append(fields, pgconn.FieldDescription{
+			Name:        rel.Columns[i].Name,
+			DataTypeOID: rel.Columns[i].DataType,
+		})
+
+		v, err := a.decodeColumnValue(rel.Columns[i], tuple.Data)
+		if err != nil {
+			return nil, err
+		}
+
+		values = append(values, v)
+	}
+
+	return a.Extract(rel.RelationName, fields, values)
+}
+
+func (a *avroExtractor) Extract(name string, fields []pgconn.FieldDescription, values []any) (avro.Schema, error) {
+	var avroFields []*avro.Field
+
+	for i, f := range fields {
+		t, ok := a.pgMap.TypeForOID(f.DataTypeOID)
+		if !ok {
+			return nil, fmt.Errorf("field %q with OID %d cannot be resolved", f.Name, f.DataTypeOID)
+		}
+
+		s, err := a.extractType(t, values[i])
+		if err != nil {
+			return nil, err
+		}
+
+		af, err := avro.NewField(f.Name, s)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create avro field %q: %w", f.Name, err)
+		}
+
+		avroFields = append(avroFields, af)
+	}
+
+	slices.SortFunc(avroFields, func(a, b *avro.Field) int {
+		return cmp.Compare(a.Name(), b.Name())
+	})
+
+	sch, err := avro.NewRecordSchema(name, avroNS, avroFields)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create avro schema: %w", err)
+	}
+
+	return sch, nil
+}
+
+func (a *avroExtractor) extractType(t *pgtype.Type, val any) (avro.Schema, error) {
+	if ps, ok := a.avroMap[t.Name]; ok {
+		return ps, nil
+	}
+
+	switch tt := val.(type) {
+	case pgtype.Numeric:
+		// N.B.: Default to 38 positions and pick the exponent as the scale.
+		fs, err := avro.NewFixedSchema(string(avro.Decimal), "",
+			38,
+			avro.NewDecimalLogicalSchema(38, int(math.Abs(float64(tt.Exp)))),
+		)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create avro.FixedSchema: %w", err)
+		}
+		return fs, nil
+	default:
+		return nil, fmt.Errorf("cannot resolve field %q of type %T", t.Name, tt)
```

Comment:
I believe that we're intentionally trying to focus only on certain types in this PR, but I just wanted to confirm that we're fine with this list of types.

Reply:
For now yea, this default here to remind us harshly if we skip anything.